Editor styles: delete duplicate backwards compat CSS custom properties

ramonjd opened the pull request:
## What and how? And why?

Fixes https://github.com/WordPress/gutenberg/issues/59701 https://github.com/WordPress/gutenberg/issues/56293

----- 

Exploring solutions to:

- https://github.com/WordPress/gutenberg/issues/59701
- https://github.com/WordPress/gutenberg/issues/56293

Alternative to:

- https://github.com/WordPress/gutenberg/pull/60398

---- 

Since https://github.com/WordPress/gutenberg/pull/42084, theme CSS custom properties reside in `:root`, and not `body`. 

`:root`, however, has less specificity than `body` and won't overwrite some low-specificity backwards compat CSS custom properties.

This PR deals exclusively with backwards compat CSS custom properties in the block editor:

```css
:where(.editor-styles-wrapper) {
	--wp--preset--font-size--normal: 16px;
	--wp--preset--font-size--huge: 42px;
}
```

This is a test to delete these properties as they are duplicates of those on [common.scss](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L7), which appear both in the top and editor frames. 

The `:root` selector is more appropriate as it reduces specificity of backwards compatibility rules using `:where()` so they do not override any theme rules in `:root`, and it's compatible with the hierarchy on the frontend:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/common.scss#L7-L15

## Testing Instructions

Go ahead and define some theme.json font sizes using the "normal" and "huge" slugs:

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"typography": {
			"fluid": true,
			"fontSizes": [
				{
					"fluid": {
						"min": ".888rem",
						"max": "1rem"
					},
					"name": "Normal",
					"size": "1rem",
					"slug": "normal"
				},
				{
					"fluid": {
						"min": "1.333rem",
						"max": "1.777rem"
					},
					"name": "Huge",
					"size": "3rem",
					"slug": "huge"
				}
			]
		}
	}
}
```

Now apply these font size presets to text elements:

```html
<!-- wp:paragraph {"fontSize":"normal"} -->
<p class="has-normal-font-size">Normal</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"fontSize":"huge"} -->
<p class="has-huge-font-size">Huge</p>
<!-- /wp:paragraph -->
```

In the editor, ensure that the theme.json preset values for `--wp--preset--font-size--normal` and `--wp--preset--font-size--huge` overwrite the backwards compat values.

To test that the backwards compat values still kick in, you can delete one or all of the presets from your theme.json and then refresh the editor.

ramonjd committed:
This is a test to move editor backwards compatibility CSS properties to the :root selector, output in the block editor iframe.

Doing this effectively reduces specificity of backwards compatibility rules using :where() so they do not override any theme rules in :root, and it's compatible with the hierarchy on the frontend in packages/block-library/src/common.scss

ramonjd committed:
Added ref to PR

ramonjd commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -254,6 +254,14 @@ function Iframe( {
+				This CSS Custom Properties aren't used anymore as defaults, but we still need to keep them for backward compatibility.
+				See: https://github.com/WordPress/gutenberg/pull/37381
+			 */
+			:root {
```
I could see this being abstracted into a filter or function so we could house any backwards compat CSS properties there. 

For now, it's a quick fix.

t-hamano commented:
Thanks for the PR!

I may be missing something, but can't we just delete [this](https://github.com/WordPress/gutenberg/pull/60400/files#diff-b24dba5ec15fc1a8f840373fbd5dcfd35c2e2e593eae75a8c1cf3ee256c1d5f2L67-L73) without injecting backward compatible styles using `:root` selectors into the iframe?

My understanding is that the CSS defined in [common.css](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L11-L14) for backward compatibility is built as `block-library/style.css` and enqueued to both the frontend and the editor. And if the editor is iframed, this style will be automatically injected into the iframe as well 🤔

ramonjd commented:
> My understanding is that the CSS defined in [common.css](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L11-L14) for backward compatibility is built as block-library/style.css and enqueued to both the frontend and the editor. And if the editor is iframed, this style will be automatically injected into the iframe as well

Thanks for double checking here. My motivation was to limit the rules to the iframe - I didn't do much digging, but I'll experiment a bit more and get back to you.

🤞🏻 it's what you say - it'd be a much cleaner approach!

ramonjd commented:
> My understanding is that the CSS defined in [common.css](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L11-L14) for backward compatibility is built as block-library/style.css and enqueued to both the frontend and the editor. And if the editor is iframed, this style will be automatically injected into the iframe as well 🤔

Yeah, thanks for pointing that out 🤔 

The styles are effectively doubled up in the editor

In the top frame (window) I'm seeing both common and editor styles. Same in the editor iframe.

It make me wonder why they're duplicated in editor.scss at all. Maybe for historical reasons that no longer apply?

Deleting these duplicate rules in editor.scss seems like the way to go on the surface. I'll do that, but it would be good to double check with @oandregal just in case

ramonjd committed:
The font size backwards compat CSS rules and properties are duplicated in common.scss.
It seems safe to delete the editor ones, which have a higher specificity and conflict with theme.json css properties with the same name.

ramonjd committed:
reinstate other

ramonjd commented on packages/block-library/src/editor.scss:
```diff
@@ -64,13 +63,6 @@
-	--wp--preset--font-size--normal: 16px;
-	--wp--preset--font-size--huge: 42px;
-}
-
```
Could we also delete in this file the following rules that are duplicated in [common.scss](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L7)?

```css
:root .editor-styles-wrapper {
	@include background-colors-deprecated();
	@include foreground-colors-deprecated();
	@include gradient-colors-deprecated();
}

:where(.editor-styles-wrapper) .has-normal-font-size {
	font-size: var(--wp--preset--font-size--normal);
}

:where(.editor-styles-wrapper) .has-huge-font-size {
	font-size: var(--wp--preset--font-size--huge);
}
```

?

They are loaded in the top-level and editor frames.

t-hamano replied:
> Could we also delete in this file the following rules that are duplicated in [common.scss](https://github.com/WordPress/gutenberg/blob/965454f2e166819d7c78b1770340411643a24d48/packages/block-library/src/common.scss#L7)?

From what I've tested, removing these styles should be fine. As I remember, @ellatrix was involved in the process of injecting the document stylesheet into the iframe editor instance. Maybe she has some knowledge.

ramonjd replied:
Thanks again. I'll remove them and ask for forgiveness later :trollface:

ellatrix replied:
I have no idea. If both stylesheets are present in the iframe, it should be fine

ramonjd replied:
Thanks for checking!

ramonjd committed:
Remove styles that are already set in common.scss

ramonjd commented:
Are we good to merge this do you think @t-hamano ?

t-hamano commented:
@ramonjd Sorry for the late confirmation! I think it's okay to ship this PR, but I would appreciate it if you could give me a few days, as I would like to test it a little more to make sure everything works as expected. I'm planning to test something like the following:

- Does it work both when the editor is iframed and when it is not iframed?
- Will it work correctly even when a classic theme that does not have a theme.json, that is, a font size is added with `editor-font-sizes`  support?
- When did `common.css` get loaded into the iframe editor instance?

t-hamano commented on packages/block-library/src/editor.scss:
```diff
@@ -55,22 +55,8 @@
-
 @import "./editor-elements.scss";
 
-:root .editor-styles-wrapper {
-	@include background-colors-deprecated();
-	@include foreground-colors-deprecated();
-	@include gradient-colors-deprecated();
-}
```
This style is not relevant to the problem we are trying to solve, so we may want to keep it just in case. It might be possible to remove it in the future. What do you think?

ramonjd replied:
Good call. I'll add it back.

t-hamano reviewed (approved):
LGTM!

As I investigated in [this comment](https://github.com/WordPress/gutenberg/issues/59701#issuecomment-2053387301), the `style(.min).css` in which the backward compatible styles were defined was already loaded in the editor, and I believe that the backward compatible styles defined in `editor(.min).css` should be safe to remove. This PR prefers styles defined in theme.json whether the editor is iframed or not.

Below is an example of a font size with a `huge` slug:

### Non-iframed Post Editor

```css
/* Inline style */
.editor-styles-wrapper {
    --wp--preset--font-size--huge: clamp(1.333rem, 1.333rem + ((1vw - 0.2rem) * 0.555), 1.777rem)
}
:root {
    --wp--preset--font-size--huge: clamp(1.333rem, 1.333rem + ((1vw - 0.2rem) * 0.555), 1.777rem);
}
/* style.css */
:root {
    --wp--preset--font-size--huge: 42px;
}
```

### Iframed Post Editor and Site Editor

```
/* Inline style */
:root {
    --wp--preset--font-size--huge: clamp(1.333rem, 1.333rem + ((1vw - 0.2rem) * 0.555), 1.777rem);
}
/* style.css */
:root {
    --wp--preset--font-size--huge: 42px;
}
```

Additionally, I tested the following:

- ✅ If font sizes with `huge` and  `normal` slugs are not defined in theme.json, the default fallback value (16px/42px) will be applied.
- ✅ Works correctly with Twenty Twenty-One. This means that even if font sizes with `huge` and `fuga` slug are defined via the `editor-font-sizes` theme support rather than in theme.json, that value will take precedence.

ramonjd commented:
@t-hamano Thanks so much for the thorough testing. Really a big help.

ramonjd committed:
Revert removal of duplicated deprecated CSS, as it's not relevant to the specific issue.
